Enable Python bindings for UntypedStorage

cbalioglu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #68945

This PR enables the Python conversion functions for `Storage` (specifically `UntypedStorage`) and also cleans up some remnants of the deprecated typed storages from `DynamicTypes.cpp`.

Differential Revision: [D32676505](https://our.internmc.facebook.com/intern/diff/D32676505/)

cc @albanD @zou3519

cbalioglu committed:
Enable Python bindings for UntypedStorage

This PR enables the Python conversion functions for `Storage` (specifically `UntypedStorage`) and also cleans up some remnants of the deprecated typed storages from `DynamicTypes.cpp`.

Differential Revision: [D32676505](https://our.internmc.facebook.com/intern/diff/D32676505/)

[ghstack-poisoned]

cbalioglu committed:
Update on "Enable Python bindings for UntypedStorage"


This PR enables the Python conversion functions for `Storage` (specifically `UntypedStorage`) and also cleans up some remnants of the deprecated typed storages from `DynamicTypes.cpp`.

Differential Revision: [D32676505](https://our.internmc.facebook.com/intern/diff/D32676505/)

[ghstack-poisoned]

zou3519 commented on tools/autograd/templates/python_variable_methods.cpp:
```diff
@@ -920,7 +920,7 @@ static PyObject * THPVariable_storage(PyObject* self, PyObject* arg)
   }
   auto& self_ = THPVariable_Unpack(self);
-  return createPyObject(self_.storage(), self_.dtype());
+  return createPyObject(self_.storage());
```
I'm not very familiar with the storage code. What does this PR enable? Is it allowing functions to return UntypedStorage to Python?

cbalioglu replied:
Yes, this is also the root cause of #66444. `__torch_dispatch__` fails for the `set_` operator because if fails to convert `UntypedStorage` to a Python object.

zou3519 replied:
Nice, thanks for the summary. Would the logic in https://github.com/pytorch/pytorch/issues/66444 be good for a test case?

albanD commented:
@kurtamohler would you have some time to check this?

albanD reviewed (commented):
Is there a test we can add to make sure we don't break this in the future?

cbalioglu commented:
> Is there a test we can add to make sure we don't break this in the future?

Makes sense. Let me add one based on #66444.

cbalioglu committed:
Update on "Enable Python bindings for UntypedStorage"


This PR enables the Python conversion functions for `Storage` (specifically `UntypedStorage`) and also cleans up some remnants of the deprecated typed storages from `DynamicTypes.cpp`.

Differential Revision: [D32676505](https://our.internmc.facebook.com/intern/diff/D32676505/)

cc albanD zou3519

[ghstack-poisoned]

albanD commented on test/test_python_dispatch.py:
```diff
@@ -522,5 +522,12 @@ def wrap(e):
+    def test_storage_can_be_converted_to_python_object(self):
+        with enable_python_mode(LoggingTensor):
+            s = torch.Storage()
+            z = LoggingTensor(torch.empty([]))
```
Does z needs to be a LoggingTensor here since we are using a mode?

cbalioglu replied:
Technically, as of today, the test would also work with a plain `z = torch.empty([])` statement. However, by design `enable_python_mode()` is meant to be used along with tensors of the type specified as its argument. So just wanted to make sure I follow that.

albanD replied:
Ok!

albanD reviewed (approved):
Thanks for adding the test!
Minor question only.
